### ISSUE TITLE
Unable to build a charm using the dump plugin

### ISSUE THREAD

**Issue by wolsen** [canonical/charmcraft#590]
When attempting to use the dump plugin for a classic charm, I get an error indicating that validation failed due to extra fields that are not permitted.
```
$ charmcraft build
Packing charm 'ceph-mon_ubuntu-20.04-amd64.charm'...
Part 'charm' validation failed.
- extra field 'charm-entrypoint' not permitted
- extra field 'charm-requirements' not permitted
Review part 'charm' and make sure it's correct. (full execution logs in '/tmp/charmcraft.log')
Failed to build charm for bases index '0'. (full execution logs in '/home/wolsen/snap/charmcraft/common/charmcraft-log-_k6m7d44')
```

My charmcraft.yaml file looks as follows:

```yaml
type: charm

parts:
  charm:
    plugin: dump
    source: .
    prime:
      - actions/*
      - files/*
      - hooks/*
      - lib/*
      - templates/*
      - actions.yaml
      - config.yaml
      - copyright
      - hardening.yaml
      - icon.svg
      - LICENSE
      - Makefile
      - metadata.yaml
      - README.md
    build-packages:
      - python3-dev


bases:
  - name: ubuntu
    channel: "20.04"
    architectures:
      - amd64
```

If interested, the classic charm with this change is [in this repository](https://github.com/wolsen/charm-ceph-mon/tree/charmcraft) (charmcraft branch).

Relevant part from the logs (below) appears to indicate that while the plugin did change to dump, the charm-requirements and charm-entrypoints are left behind with the default values. However, when validating the dump plugin does not define those as valid parameters, which causes the build to fail:
```
2021-11-04 18:58:06,608  charmcraft.providers                     DEBUG    Logs captured from managed instance:
mm 2021-11-05 01:58:06,147  charmcraft.guard                         DEBUG    Starting charmcraft version 1.2.1
mm 2021-11-05 01:58:06,148  charmcraft.main                          DEBUG    Raw pre-parsed sysargs: args={'help': False, 'verbose': False, 'quiet': False, 'project_dir': None} filtered=['pack', '--bases-index', '0']
mm 2021-11-05 01:58:06,159  charmcraft.main                          DEBUG    General parsed sysargs: command='pack' args=['--bases-index', '0']
mm 2021-11-05 01:58:06,161  charmcraft.main                          DEBUG    Command parsed sysargs: Namespace(bases_index=[0], destructive_mode=False, entrypoint=None, force=False, requirement=None)
mm 2021-11-05 01:58:06,162  charmcraft.commands.pack                 DEBUG    Working arguments: {'from': PosixPath('/root/project'), 'destructive_mode': False, 'entrypoint': None, 'requirement': [], 'bases_indices': None, 'force': False}
mm 2021-11-05 01:58:06,162  charmcraft.metadata                      DEBUG    Parsing '/root/project/metadata.yaml'
mm 2021-11-05 01:58:06,171  charmcraft.commands.build                DEBUG    Building for 'bases[0]' as host matches 'build-on[0]'.
mm 2021-11-05 01:58:06,171  charmcraft.commands.build                DEBUG    Building charm in '/root/project/build'
mm 2021-11-05 01:58:06,171  charmcraft.commands.build                DEBUG    Parts definition: {'charm': {'plugin': 'dump', 'source': '/root/project', 'prime': ['actions/*', 'files/*', 'hooks/*', 'lib/*', 'templates/*', 'actions.yaml', 'config.yaml', 'copyright', 'hardening.yaml', 'icon.svg', 'LICENSE', 'Makefile', 'metadata.yaml', 'README.md', 'src', 'venv', 'metadata.yaml', 'dispatch', 'hooks', 'config.yaml', 'actions.yaml', 'templates', 'lib', 'LICENSE', 'icon.svg', 'README.md'], 'build-packages': ['python3-dev'], 'charm-requirements': ['requirements.txt'], 'charm-entrypoint': 'src/charm.py'}}
mm 2021-11-05 01:58:06,172  craft_parts.utils.os_utils               DEBUG    is_snap: True, SNAP_NAME set to charmcraft
mm 2021-11-05 01:58:06,172  craft_parts.infos                        DEBUG    Setting target machine to x86_64
mm 2021-11-05 01:58:06,172  charmcraft                               ERROR    Part 'charm' validation failed.
- extra field 'charm-entrypoint' not permitted
- extra field 'charm-requirements' not permitted
Review part 'charm' and make sure it's correct. (full execution logs in '/tmp/charmcraft.log')
```

**Comment by wolsen** [comment]
A bit of a hacky way to work around this is to leverage the override-{build,stage,prime} as follows:

```yaml
type: charm

parts:
  charm:
    override-build: |
      echo "Classic charm, nothing to build"
    override-stage: |
      cp -r $CHARMCRAFT_PART_BUILD/* $CHARMCRAFT_STAGE/
    override-prime: |
      cp -r $CHARMCRAFT_STAGE/* .


bases:
  - name: ubuntu
    channel: "20.04"
    architectures:
      - amd64
```

**Comment by cmatsuoka** [comment]
We're discussing a few changes to make part handling more uniform and predictable, which would allow something similar to your first example. Stay tuned.

**Comment by cmatsuoka** [comment]
This should be addressed by PR #596.
